### PR TITLE
feat(dw): partial auto aggregation function, specifically, deduplication

### PR DESCRIPTION
**Pull request by GuangMingYouBei**
```
const dataFrame = new DataFrame(inputData);
const deduplicatedData = new AutoAggregation(dataFrame).deDuplication();
```
A deduplication method is proposed to aggregate the duplicated data rows which are troublesome for visualization.

### CODE REVIEW

File: packages/data-wizard/src/dataset/aggregation/utils.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+      return [...new Set(value)].length;
+    }
+    default: {
+      return sum(value);
+    }
+  }
+};
+
+export const aggregateDataFrame = (
+  inputData: DataFrame,
```
Comment (BBSQQ) [inline packages/data-wizard/src/dataset/aggregation/utils.ts:59]: Maybe it's more appropriate to call dataframe directly?

File: packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  constructor(dataFrame: DataFrame) {
+    this.dataFrame = dataFrame;
+  }
+
+  deDuplication() {
+    const dimensions = this.dataFrame.axes[1].filter((name) => {
+      const index = this.dataFrame.axes[1].indexOf(name);
+      return !this.dataFrame.info()[index].levelOfMeasurements.includes('Interval');
+    });
```
Comment (chenluli) [inline packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts:16]: `includes('Interval')` 可以加注释解释下思路？'Interval' 是不是可以写成常量

File: packages/data-wizard/__tests__/dataset/aggregation.test.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  const inputData = [
+    { gender: 'male', score: 99 },
+    { gender: 'male', score: 98 },
+    { gender: 'male', score: 97 },
+    { gender: 'female', score: 99 },
+    { gender: 'female', score: 98 },
+    { gender: 'female', score: 92 },
+  ];
+  const dataFrame = new DataFrame(inputData);
+  expect(aggregateDataFrame(dataFrame, ['gender'], ['score']).data[0][1]).toStrictEqual(294);
```
Comment (neoddish) [inline packages/data-wizard/__tests__/dataset/aggregation.test.ts:15]: 这里可以直接输出结果对比，也不复杂，不用取data[0][1]。能在 test case 里看到结果不是很好么

File: packages/data-wizard/src/dataset/aggregation/utils.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
+import { meanBy, sumBy, minBy, maxBy, groupBy, sum, mean, min, max } from '../../statistics';
+import { DataFrame } from '..';
+
+type Datum = Record<string, string | number>[];
+
+const getAggregatedMeasure = (value: any[], measure: string, agg?: string): number => {
+  switch (agg) {
```
Comment (chenluli) [inline packages/data-wizard/src/dataset/aggregation/utils.ts:7]: 参数：agg 应该有具体的联合类型？measure 是 measureId ？ 可以指明一下

File: packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { DataFrame } from '..';
+
+import { aggregateDataFrame } from './utils';
+
+export default class AutoAggregation {
+  dataFrame: DataFrame;
```
Comment (ArranZeyuWang) [inline packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts:6]: 这里只有 autoAgg 一个方法，不如直接加在 df 下面

File: packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { DataFrame } from '..';
+
+import { aggregateDataFrame } from './utils';
+
+export default class AutoAggregation {
```
Comment (neoddish) [inline packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts:5]: AutoAggregation 怎么看都是一个 function，需要设计成类吗？设计成 class 的必要性是？

File: packages/data-wizard/src/dataset/aggregation/utils.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    // terminal condition of recursion
+    const obj = {};
+    for (let i = 0; i < dimensions.length; i += 1) {
+      const dimId = dimensions[i];
+      obj[dimId] = sequence[0][dimId];
+    }
+
+    for (let j = 0; j < measures.length; j += 1) {
+      const meaId = measures[j];
+      const agg = Object.prototype.hasOwnProperty.call(aggMethod, meaId) ? aggMethod[meaId] : 'sum';
```
Comment (BBSQQ) [inline packages/data-wizard/src/dataset/aggregation/utils.ts:130]:
```js
const agg = aggMethod?.[meaId] || 'sum';
```

File: packages/data-wizard/src/dataset/aggregation/utils.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
+import { meanBy, sumBy, minBy, maxBy, groupBy, sum, mean, min, max } from '../../statistics';
+import { DataFrame } from '..';
+
+type Datum = Record<string, string | number>[];
+
+const getAggregatedMeasure = (value: any[], measure: string, agg?: string): number => {
```
Comment (neoddish) [inline packages/data-wizard/src/dataset/aggregation/utils.ts:6]: 缺少 jsdoc。这个方法具体是做什么的？

/**
* XXXXX
*/

File: packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (GuangMingYouBei) [reply packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts:5]: because more functions will be proposed, later on, deduplication is just a small functionality.

File: packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import { DataFrame } from '..';
+
+import { aggregateDataFrame } from './utils';
+
```
[the remark this answers is not in the document]
Comment (GuangMingYouBei) [reply packages/data-wizard/src/dataset/aggregation/auto-aggregation.ts:4]: 因为后期要接入更多方法。

File: packages/data-wizard/src/dataset/aggregation/utils.ts [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+    if (!Object.prototype.hasOwnProperty.call(tempResult, tempKey)) {
+      tempResult[tempKey] = {};
+      for (let i = 0; i < dimensionsIndex.length; i += 1) {
+        const dimColId = dimensionsIndex[i];
+        tempResult[tempKey][dimColId] = item[dimColId];
+      }
+      for (let j = 0; j < measuresIndex.length; j += 1) {
+        const agg =
+          aggMethod && Object.prototype.hasOwnProperty.call(aggMethod, measures[j]) ? aggMethod[measures[j]] : 'sum';
```
Comment (ArranZeyuWang) [inline packages/data-wizard/src/dataset/aggregation/utils.ts:84]: aggMethod?.[measures[j]] ?? 'sum'

File: packages/data-wizard/src/dataset/aggregation/utils.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+};
+
+export const aggregateDataFrame = (
+  inputData: DataFrame,
+  dimensions: string[] | any[],
+  measures: string[] | any[],
+  aggMethod?: Record<string, string>
+): DataFrame => {
+  const tempResult = {};
+  const dimensionsIndex = dimensions.map((item) => inputData.axes[1].indexOf(item));
```
Comment (neoddish) [inline packages/data-wizard/src/dataset/aggregation/utils.ts:65]: 各种取 .axes[1] 的可读性太差了，建议抽象出一个 help func

File: packages/data-wizard/src/dataset/aggregation/utils.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
+import { meanBy, sumBy, minBy, maxBy, groupBy, sum, mean, min, max } from '../../statistics';
+import { DataFrame } from '..';
+
+type Datum = Record<string, string | number>[];
```
[the remark this answers is not in the document]
Comment (GuangMingYouBei) [reply packages/data-wizard/src/dataset/aggregation/utils.ts:4]: dw的使用逻辑是 万物转 dataframe，所以不应该 export 更多数据类型？

File: packages/data-wizard/src/dataset/aggregation/utils.ts [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+  const dimensionsIndex = dimensions.map((item) => inputData.axes[1].indexOf(item));
+  const measuresIndex = measures.map((item) => inputData.axes[1].indexOf(item));
+
+  inputData.data.forEach((item) => {
+    const tempKeyList = [];
+    for (let i = 0; i < dimensionsIndex.length; i += 1) {
+      const dimColId = dimensionsIndex[i];
+      tempKeyList.push(item[dimColId]);
+    }
+    const tempKey = tempKeyList.join('-');
```
Comment (neoddish) [inline packages/data-wizard/src/dataset/aggregation/utils.ts:74]: 如果 tempKeyLIst 的元素、item[dimColId] 内本身就带”-“怎么办？怎么处理

File: packages/data-wizard/src/dataset/aggregation/utils.ts [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    Object.keys(temp).forEach((index) => {
+      datum[inputData.axes[1][index]] = temp[index];
+      return null;
+    });
+    return datum;
+  });
+  return new DataFrame(Object.values(result));
+};
+
+export const sliceAndAggregate = (
```
Comment (chenluli) [inline packages/data-wizard/src/dataset/aggregation/utils.ts:113]: 建议加一下函数说明和参数说明注释

File: packages/data-wizard/src/dataset/aggregation/utils.ts (hunk 10 above)
Comment (GuangMingYouBei) [reply packages/data-wizard/src/dataset/aggregation/utils.ts:84]: 羽熙提供了优雅的写法：
const agg = aggMethod?.[meaId] || 'sum';

File: packages/data-wizard/src/dataset/aggregation/utils.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (ArranZeyuWang) [reply packages/data-wizard/src/dataset/aggregation/utils.ts:4]: > Datum 很多地方都有用到，不如直接在 dw 最外层 export 出来

cc @pddpd 🤔

File: packages/data-wizard/src/dataset/aggregation/utils.ts (hunk 7 above)
Comment (GuangMingYouBei) [reply packages/data-wizard/src/dataset/aggregation/utils.ts:130]: get

File: packages/data-wizard/__tests__/dataset/aggregation.test.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
+import DataFrame from '../../src/dataset/data-frame';
+import { aggregateDataFrame } from '../../src/dataset/aggregation/utils';
+import AutoAggregation from '../../src/dataset/aggregation/auto-aggregation';
+
+test('aggregate', () => {
```
Comment (neoddish) [inline packages/data-wizard/__tests__/dataset/aggregation.test.ts:5]: 测试用例太少了！只是为了写个 demo，根本不叫测试。

各种 edge case 要考虑呀

File: packages/data-wizard/src/dataset/aggregation/utils.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (ArranZeyuWang) [reply packages/data-wizard/src/dataset/aggregation/utils.ts:4]: > dw的使用逻辑是 万物转 dataframe，所以不应该 export 更多数据类型？

也可，LI 有导出 Datum，我经常 import LI 的用，显得比较奇怪
